### DISCUSSION TITLE
Day 9 · Overreliance on reranker (No.5, No.6)

### DISCUSSION THREAD

**Discussion by onestardao** [onestardao/WFGY#34] [General]
# Day 9 · Overreliance on reranker (No.5, No.6)

> Symptom
> a strong reranker seems to “fix” answers, but offline gains vanish in real use. the base retriever stays sick. coverage is low, geometry is off, and the reranker only hides the problem until it fails on new queries.

**ProblemMap**

* **No.5 Semantic ≠ Embedding**
  base-space geometry or policy is wrong. metric, normalization, chunk contract, or tokenizer issues.
* **No.6 Logic Collapse & Recovery**
  synthesis over-interprets thin evidence. reranker sometimes masks the collapse instead of forcing a recovery step.

---

## What this usually looks like

* with reranker on, a few showcase queries look great. switch it off and recall collapses
* top-k from the raw retriever rarely contains the true span. reranker promotes “close but not it”
* high MRR in offline tables, yet citations do not match spans when humans read them
* small paraphrase flips the outcome. with reranker off, answers alternate across runs
* cosine vs L2 confusion. half the store normalized, half not. reranker can not fix metric mismatch
* chunks are too large or mis-aligned. reranker picks generic summaries over claim-aligned spans

---

## 60-sec quick test

1. **ablation**
   run the same question twice
   a) retriever only
   b) retriever + reranker

2. **measure**

   * coverage of the target section in top-k
   * ΔS(question, retrieved)
   * cite count per atomic claim

3. **label**

   * if coverage is low without reranker and skyrockets only with reranker, suspect **No.5**
   * if coverage is ok but freeform prose drifts, suspect **No.6**

4. **stability**
   paraphrase the question three ways. if labels or answers alternate, your chain is unstable and the reranker is hiding it.

Rules of thumb

* coverage of target section **≥ 0.70** before reranking
* ΔS(question, retrieved) **≤ 0.45** for stable chains
* at least one valid citation **per atomic claim**, even with reranker on

---

## Diagnose checklist

* metric mismatch. IP index over already normalized vectors or L2 over non-normalized
* mixed shards. part cosine style, part L2, part whitened, part not
* chunking to embedding contract is missing. headers and boilerplate dominate top-k
* vectorstore fragmentation. near duplicates split evidence across ids
* citation-first template not enforced. synthesis fills gaps instead of pausing
* reranker trained or configured to prefer generic summaries over tight spans
* eval set is tiny and overfits to reranker behavior

---

## Minimal fix

**Goal** repair the base space first, then keep reranking as a light, auditable layer.

1. **Align metric and normalization**

```python
# cosine style retrieval: L2-normalize both sides, then use L2 or IP consistently
from sklearn.preprocessing import normalize
Z = normalize(Z, axis=1).astype("float32")
q = normalize(q, axis=1).astype("float32")
# document which metric the index uses, and keep it consistent everywhere
```

2. **Enforce chunk → embed contract**

   * mask boilerplate and headers
   * keep chunk sizes consistent with your model’s token window
   * add stable `snippet_id, section_id, offsets, tokens`

3. **Coverage gate before rerank**
   do not let reranker run if base top-k has coverage < 0.70. return a **bridge** asking for better retrieval or more context.

4. **Citation-first schema**
   lock cite then explain. fail fast if claims lack in-scope snippet ids.

5. **Rerank for span alignment only**
   keep reranker weights modest. prefer claim-aligned spans, not generic summaries. record the reranker score alongside citations.

---

## Hard fixes when minimal is not enough

* rebuild the index from clean embeddings. one metric, one normalization policy
* re-train IVF or PQ codebooks on a large, deduped, diverse sample
* de-duplicate near duplicates and collapse boilerplate before embedding
* add a lexical or BM25 leg for sparse match, then fuse with a simple rule
* add a light cross-encoder rerank, but only after the base candidate set is healthy

---

## WFGY guardrails you can enable

* **Retrieval Traceability**. log `snippet_id, section_id, offsets, tokens, rerank_score`
* **Regression gate**. block publish if coverage < 0.70 or ΔS > 0.45 even when reranker says ok
* **Bridge step** for thin evidence. state what is missing, ask for the next snippet id
* **Variance clamp**. measure λ across three paraphrases. reject divergent chains

> in forums, label this class as **No.5** when base geometry is wrong. add **No.6** if synthesis still drifts after coverage is fixed.

---

## Tiny scripts you can paste

**base vs rerank lift check**

```python
def lift_at_k(gt_ids, base_ids, rerank_ids, k=10):
    base_hit = int(any(i in gt_ids for i in base_ids[:k]))
    rr_hit   = int(any(i in gt_ids for i in rerank_ids[:k]))
    return {"base_hit": base_hit, "rerank_hit": rr_hit, "lift": rr_hit - base_hit}
```

**neighbor-overlap sanity**

```python
def overlap_at_k(a_ids, b_ids, k=20):
    a, b = set(a_ids[:k]), set(b_ids[:k])
    return len(a & b) / float(k)  # healthy spaces sit well below 0.35
```

**coverage gate**

```python
def coverage_ok(retrieved_ids, target_ids, k=10, threshold=0.70):
    hits = sum(1 for i in retrieved_ids[:k] if i in target_ids)
    return (hits / float(min(k, len(target_ids)))) >= threshold
```

---

## Acceptance checks before you call it fixed

* base top-k contains the target section with coverage **≥ 0.70**
* ΔS(question, retrieved) **≤ 0.45** across three paraphrases
* citations exist per claim and match in-scope snippets
* reranker improves ranking without hiding base failures. lift is positive yet not required for correctness

---

## TL;DR

rerankers are great polish, not a crutch. if the base space is wrong, a reranker only hides the pain. fix metric and normalization, fix chunk contracts, demand coverage and citations, then add a light rerank for span alignment. call it **No.5** and **No.6** when synthesis still drifts.

---

[https://github.com/onestardao/WFGY/blob/main/ProblemMap/article/README.md](https://github.com/onestardao/WFGY/blob/main/ProblemMap/article/README.md)
